### ISSUE TITLE
[m_spanningtree] Servers not linking when server 1 has AAAA record but link attempts between both are in IPv4

### ISSUE THREAD

**Issue by mahriman** [inspircd/inspircd#1811]
If server 1's hostname has an AAAA record, server 2 (bound to IPv4 address and not speaking IPv6) will not link with server 1 and return error 
`LINK: Connection to 'irc.example.org' failed with error: Invalid argument`, even when server 1 has an A-record.
If server 1 tries to link with server 2 (whose domain name only has an A record) (and server 1's `bind` value in `<link>` is server 1's IPv4-address), server 1 will be met with a `Connection reset by peer` when trying to link, while server 2's logs will indicate that it cannot find a matching `<link>` for server 1's ipv4 address, even though server 1 has an A record that points to the IP address the connect is happening from.

It would seem that at least in version 2.0.27 (server 2 in this example), inspircd receives the AAAA record and tries to connect to the IPv6 address with IPv4 (causing the `failed with error: Invalid argument` error), while failing to match an incoming link's IPv4 address with the A-record of the hostname in the `<link>` `ipaddr` field if that hostname also has an AAAA record.

Switching the `ipaddr` field of server 2's `<link>` block to a hostname with only an A record for server 1 allows linking without errors in both directions..

I guess the workaround of providing an non-AAAA-record hostname for the `<link` `ipaddr` attribute is viable, however I think that inspircd should handle the problem more gracefully (e.g. at least at debug log level state each link block server 2 is trying to match the connecting server 1's ip with, and the address(es) it resolves server 1's `<link ipaddr` hostname to and uses for comparison).

**Server 1**: InspIRCd-3.7.0-FreeBSD-3.7.0 (ports)
**Server 2**: 2.0.27-1+deb10u1

**Comment by SadieCat** [comment]
Does it work if you link the other way around? I think I fixed this in v3 already and v2 is only receiving security fixes at this point (it goes EOL in two~ months).

**Comment by mahriman** [comment]
> Does it work if you link the other way around? I think I fixed this in v3 already and v2 is only receiving security fixes at this point (it goes EOL in two~ months).

Link cannot be initiated from either server and results in two different problems/errrors (server 1 initiating link: server 2 is unable to match IPv4 address with domain resolution of server 1. Server 2 initiating the link: server 2 is unable to make link connection to server 2 because it seems to try to connect to an IPv6 address over IPv4)

If you believe it's fixed in v3, forget this report :)

**Comment by mahriman** [comment]
Re-opening as this issue somewhat persists but in a slightly different manner:

**Server 1:**  InspIRCd-3.11.0-FreeBSD-3.11.0 (ports)
**Server 2:** 3.8.1

Server 1 (let's call it _bowser.server1domain.org_) has both an AAAA and an A record. Server 1 also has an alternative domain (_server1-2nddomain.org_) with only an A record.
Server 2 (let's call it _irc.server2domain.org_) has only an A record.

Both servers have their <link> blocks _ipaddr_ attribute configured with host names. 

When Server 2 tries to connect to Server 1, all is fine and dandy.
But if Server 2's <link> _ipaddr_ attribute is set to Server 1's IPv6-enabled domain (with both A and AAAA records), Server 1's oper link message will tell it's a "Connection reset by peer" problem:
`00:10:00 [KKK-net] !bowser.server1domain.org *** LINK: AUTOCONNECT: Auto-connecting server irc.server2domain.org`
`00:10:00 [KKK-net] !bowser.server1domain.org *** LINK: Connection to irc.server2domain.org[irc.server2domain.org] started.`
`00:10:00 [KKK-net] !bowser.server1domain.org *** LINK: Connection to 'irc.server2domain.org' failed with error: Connection reset by peer`
`00:10:00 [KKK-net] !bowser.server1domain.org *** LINK: Connection to 'irc.server2domain.org' failed.`

But Server 2's inspircd logs will state:
`Wed Sep 29 2021 23:52:10 LINK: Server connection from w.x.y.z denied (no link blocks with that IP address)`
`Wed Sep 29 2021 23:52:10 SOCKET: Refusing connection on a.b.c.d:7000 - Connection refused by module`
Where w.x.y.z is the correct IPv4 address of Server 1 that both bowser.server1domain.org's A record and server1-2nddomain.org's A record point to, and a.b.c.d is an internal 192.168.x.x address that server 2 is running on (port forwarding is set up, don't worry!)

If however Server 2's administrator changes the _ipaddr_ attribute of the <link> block of on Server 2 to use the second hostname - server1-2nddomain.org - Server 1 is able to initiate a linking to Server 2 as well.

- - -

So to recap: If Server 2 has a Server 1 hostname that only has an A record, both servers can initiate linking. If Server 2 has entered a hostname that has both AAAA and A records, Server 2 is able to initiate the link with Server 1 but not the other way aroud, resulting in "Connection reset by peer" according to Server 1's oper link message and logs on Server 2 stating there's no IP address matching the link even though the Server 1 hostname has a valid A record that resolves to the same IP address inspircd says it can't find a matching link record for.

**Comment by SadieCat** [comment]
I think I've splatted this bug now. It'll be in the next release which comes out next Friday.
